addon/addon/models/animation-participant.ts

```diff
@@ -32,39 +32,115 @@ export class AnimationParticipantManager {
+        displacedDOMRefs.add(node)
+      );
+      if (
+        participant.uiState.detached &&
+        (!participant.uiState.detached.animation ||
+          participant.uiState.detached.animation.playState === 'finished' ||
+          participant.uiState.detached.animation?.playState === 'idle' ||
+          participant.uiState.detached.animation?.playState === 'paused')
+      ) {
+        // TODO: when cloning is implemented, add the DOMRef here
+        unanimatedDetachedDOMRefs.add(participant.uiState.detached.DOMRef);
+      }
+    }
+
+    interface Scope {
+      lastLiveAncestor: DOMRefNode | undefined | 'ROOT';
+      isOnDetachedBranch: boolean;
+      stop: boolean;
+    }
+    let deleted: DOMRefNode[] = [];
+    let nodesToGraft = new Map<DOMRefNode, DOMRefNode | 'ROOT'>();
+    let recurse = (
+      node: DOMRefNode,
+      callback: (node: DOMRefNode, scope: Scope) => Scope,
+      scope: Scope
+    ) => {
+      let nextScope = callback(node, scope);
+      if (nextScope.stop) return;
+      for (let child of node.children) {
+        recurse(child, callback, nextScope);
+      }
+    };
+    for (let DOMRef of this.DOMRefs) {
+      recurse(
+        DOMRef,
+        (node, scope) => {
+          if (unanimatedDetachedDOMRefs.has(node)) {
+            if (node.parent) node.delete();
+            else this.DOMRefs = this.DOMRefs.filter((v) => v !== node);
+            deleted.push(node);
+            return {
+              ...scope,
+              isOnDetachedBranch: true,
+            };
+          } else if (displacedDOMRefs.has(node)) {
+            if (node.parent) node.delete();
+            else this.DOMRefs = this.DOMRefs.filter((v) => v !== node);
+            deleted.push(node);
+            displacedDOMRefs.delete(node);
+            return {
+              ...scope,
+              isOnDetachedBranch: true,
+              stop: true, // no need to iterate into the children of this node. it and its children are displaced (and hence replaced with another subtree)
+            };
+          } else if (scope.isOnDetachedBranch) {
+            nodesToGraft.set(node, scope.lastLiveAncestor ?? 'ROOT');
+            return {
+              ...scope,
+              stop: true, // no need to iterate into the children of this node, because it is currently being used (it has animations on it)
+            };
+          } else {
+            // We're still on a live branch, so return this node as the lastLiveAncestor
+            // It will become the graft target later
+            return {
+              ...scope,
+              lastLiveAncestor: node,
+            };
+          }
+        },
+        {
+          isOnDetachedBranch: false,
+          lastLiveAncestor: undefined,
+          stop: false,
+        }
+      );
+    }
+
+    for (let node of displacedDOMRefs) {
+      if (node.parent) node.delete();
+      else this.DOMRefs = this.DOMRefs.filter((v) => v !== node);
+      deleted.push(node);
+    }
```

Comment:
Anything marked ready for disposal because it's displaced results in a complete subtree removal.